internal/controller/devicediscovery_controller.go

```diff
@@ -0,0 +1,172 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/Mellanox/nic-configuration-operator/pkg/consts"
+	"github.com/Mellanox/nic-configuration-operator/pkg/host"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/Mellanox/nic-configuration-operator/api/v1alpha1"
+)
+
+var deviceDiscoveryReconcileTime = time.Minute * 5
+
+// DeviceDiscovery periodically reconciles devices on the host, creates CRs for new devices,
+// deletes CRs for absent devices, updates the CR when device's status has changed.
+type DeviceDiscovery struct {
+	client.Client
+
+	hostManager host.HostManager
+	nodeName    string
+	namespace   string
+}
+
+// Constructs a unique CR name based on the device's type and serial number
+func (d *DeviceDiscovery) getCRName(deviceType string, serialNumber string) string {
+	return strings.ToLower(d.nodeName + "-" + deviceType + "-" + serialNumber)
+}
+
+func setInitialsConditionsForDevice(device *v1alpha1.NicDevice) {
+	condition := metav1.Condition{
+		Type:    consts.ConfigUpdateInProgressCondition,
+		Status:  metav1.ConditionFalse,
+		Reason:  consts.DeviceConfigSpecEmptyReason,
+		Message: "Device configuration spec is empty, cannot update configuration",
+	}
+	meta.SetStatusCondition(&device.Status.Conditions, condition)
+}
+
+// reconcile reconciles the devices on the host by comparing the observed devices with the existing NicDevice custom resources (CRs).
+// It deletes CRs that do not represent observed devices, updates the CRs if the status of the device changes,
+// and creates new CRs for devices that do not have a CR representation.
+func (d *DeviceDiscovery) reconcile(ctx context.Context) error {
+	observedDevices := d.hostManager.DiscoverNicDevices()
+
+	list := &v1alpha1.NicDeviceList{}
+
+	selectorFields := fields.OneTermEqualSelector("status.node", d.nodeName)
+
+	err := d.Client.List(ctx, list, &client.ListOptions{FieldSelector: selectorFields})
+	if err != nil {
+		log.Log.Error(err, "failed to list NicDevice CRs")
+		return err
+	}
+
+	for _, nicDeviceCR := range list.Items {
+		observedDeviceStatus, exists := observedDevices[nicDeviceCR.Status.SerialNumber]
+
+		if !exists {
+			// Need to delete this CR, it doesn't represent the observedDevice on host anymore
+			err = d.Client.Delete(ctx, &nicDeviceCR)
```

Comment:
One thing im worried about here: 
in DiscoverNicDevices its enough for any command to fail and the device is skipped thus not present in observedDevices which will trigger deletion of NICDevices (loosing their condition, state etc).

perhaps we should be strict in DiscoverNicDevices ? that is:

1. if we fail to list PCI devices -> we fail and retry
2. if we fail to get any information on a mlnx device -> we fail and retry
3. retry in this case should be sooner than 5 min imo.

WDYT ?

Reply:
with current logic, device will disappear for at least 5 minutes, then get recreated and reconfigured. (loosing its last applied state)

Reply:
One thing to notice is that if we make DiscoverNicDevices stricter, one faulty device will block all devices on that node from creating if we consistently fail to get some info about it, which could be okay as long as we report it properly.

WDYT?

Reply:
So I would prefer making DiscoverNicDevices a bit stricter and fail if we can't list PCI devices. I wouldn't fail when we can't get some device information on Mellanox devices. For example, there can be management interfaces like this:
`3b:00.2 DMA controller [0801]: Mellanox Technologies MT42822 BlueField-2 SoC Management Interface [15b3:c2d3]`
That do not support mstvpd calls. So we either make a list of device types that we want to skip (there can be others) or not fail on PN/SN/FW version/PSID retrievals.

When we do fail, though, we should requeue right away

Reply:
discovery logic should skip non network PCI device class with IB or ETH subclass see [1]
we do the this in sriov device plugin.

for what is left id expect this to work. anyway i leave it up to you to decide how to handle.

[1] https://pcisig.com/sites/default/files/files/PCI_Code-ID_r_1_11__v24_Jan_2019.pdf

Reply:
Makes sense to me. Added filtering based on the PCI device class, and now we fail when some properties of mellanox network device couldn't be retrieved.